graspy/embed/oosase.py

```diff
@@ -0,0 +1,416 @@
+    space in which to represent the graph.
+
+    References
+    ----------
+    .. [1] Sussman, D.L., Tang, M., Fishkind, D.E., Priebe, C.E.  "A
+       Consistent Adjacency Spectral Embedding for Stochastic Blockmodel Graphs,"
+       Journal of the American Statistical Association, Vol. 107(499), 2012
+    .. [2] Levin, K., Roosta-Khorasani, F., Mahoney, M. W., & Priebe, C. E. (2018). Out-of-sample 
+        extension of graph adjacency spectral embedding. PMLR: Proceedings of Machine Learning 
+        Research, 80, 2975-2984.
+    """
+
+    def __init__(
+        self,
+        n_components=None,
+        n_elbows=2,
+        algorithm="randomized",
+        n_iter=5,
+        check_lcc=True,
+        in_sample_proportion=1,
+        in_sample_idx=None,
+        connected_attempts=100,
+        semi_supervised=False,
+        random_state=None,
+    ):
+        super().__init__(
+            n_components=n_components,
+            n_elbows=n_elbows,
+            algorithm=algorithm,
+            n_iter=n_iter,
+            check_lcc=check_lcc,
+        )
+
+        if random_state is None:
+            random_state = np.random.randint(10 ** 6)
+        np.random.seed(random_state)
+
+        if in_sample_proportion <= 0 or in_sample_proportion > 1:
+            if in_sample_idx is None:
+                msg = (
+                    "must give either proportion of in sample indices or a list"
+                    + " of in sample vertices"
+                )
+                raise ValueError(msg)
+        self.connected_attempts = connected_attempts
+        self.semi_supervised = semi_supervised
+        self.in_sample_proportion = in_sample_proportion
+        self.in_sample_idx = in_sample_idx
+
+    def fit(self, graph, y=None):
+        """
+        Fit ASE model to input graph
+
+        Parameters
+        ----------
+        graph : array_like or networkx.Graph
+            Input graph to embed.
+
+        Returns
+        -------
+        self : returns an instance of self.
+        """
+
+        if isinstance(graph, np.ndarray):
+            check_array(graph)
+            graph = nx.Graph(graph).copy()
+
+        if isinstance(graph, nx.Graph):
+            if nx.is_directed(graph):
+                self._directed = True
+            else:
+                self._directed = False
+        else:
+            msg = "only arrays and networkx graphs allowed"
+            raise TypeError(msg)
+
+        self._nodes = list(graph.nodes)
+
+        if self.check_lcc:
+            if not is_fully_connected(graph):
+                msg = (
+                    "Input graph is not fully connected. Results may not"
+                    + " be optimal. You can compute the largest connected component by"
+                    + " using ``graspy.utils.get_lcc``."
+                )
+                warnings.warn(msg, UserWarning)
+
+        N = len(graph)
```

Comment:
minor gripe but n_verts would be more descriptive name

Reply:
changed `N` to `n_verts` and `n` to `n_in_sample_verts` everywhere